EU4toV2/Source/V2World/Diplomacy/Diplomacy.cpp

```diff
@@ -63,46 +68,106 @@ void V2::Diplomacy::convertDiplomacy(
-			agreement.getAgreementType() == "protectorate" || 
-			agreement.getAgreementType() == "tributary_state" || 
-			agreement.getAgreementType() == "march" || 
-			agreement.getAgreementType() == "colonial" ||
-			agreement.getAgreementType() == "colony" ||
-			agreement.getAgreementType() == "union" ||
-			agreement.getAgreementType() == "personal_union")
+		if (agreementMapper.isAgreementInDoublesiders(agreement.getAgreementType()))
+		{
+			// doublesiders are bidirectional; influence level +1, but never exceed 4
+			// They don't set military access, as it's not implied (not even for alliances).
+			if (r1.getLevel() < 4) r1.setLevel(r1.getLevel() + 1);
+			r1.increaseRelations(50);
+			if (r2.getLevel() < 4) r2.setLevel(r2.getLevel() + 1);
+			r2.increaseRelations(50);
+		}
+
+		if (agreementMapper.isAgreementInTributaries(agreement.getAgreementType()))
+		{
+			// influence level 5 - sphere, but not vassal, and military access is implied.
+			r1.setLevel(5);
+			r1.increaseRelations(75);
+			r1.setInfluence(20);
+			r1.setAccess(true);
+			r2.setAccess(true);
+		}
+
+		if (agreementMapper.isAgreementInVassals(agreement.getAgreementType()))
 		{
 			// Yeah, we don't do marches, clients or all that. Or personal unions. PUs are a second relation
 			// beside existing vassal relation specifying when vassalage ends.
-			// We assume all rulers are CK2 immortals and vassalage does not end with a specific date.
+			// However, vanilla Vic2 has PU end dates based on historical events, and we don't simulate those (for now).
 			agreement.setAgreementType("vassal");
 			r1.setLevel(5);
-			country2->second->addPrestige(-country2->second->getPrestige());
+			r1.increaseRelations(75);
+			r1.setInfluence(50);
+			// In vic2 military access through vassals is not automatic but is implied.
+			r1.setAccess(true);
+			r2.setAccess(true);
+
+			// We need to calculate devs and conglomerate devs (vassals + overlord) so that we can alter starting prestige
+			// of individual vassals. We cannot do so yet as we don't know who's alive, dead or a vassal at all.
+			if (!vassalCache[V2Tag2]) vassalCache[V2Tag2] = country2->second->getSourceCountry()->getTotalDev();
+			if (!masterCache[V2Tag1]) masterCache[V2Tag1] = country1->second->getSourceCountry()->getTotalDev();
+			masterVassals[V2Tag1].insert(V2Tag2);
 		}
 
-		// In essence we only have 3 diplomacy categories and these are it.
+		// In essence we should only recognize 3 diplomacy categories and these are it.
 		if (agreement.getAgreementType() == "alliance" || agreement.getAgreementType() == "vassal" || agreement.getAgreementType() == "guarantee")
 		{
 			// copy agreement
 			Agreement v2agreement(V2Tag1, V2Tag2, agreement.getAgreementType(), agreement.getStartDate());
 			agreements.push_back(v2agreement);
 		}
 	}
+	// Now, alter prestige of vassals.
+	for (const auto& conglomerate: masterVassals)
+	{
+		const auto& masterDev = masterCache[conglomerate.first];
```

Comment:
I'm not overly fond of the [] operator on maps. It can cause changes in the map. While the overall logic of this function makes that impossible (as conglomerate.first is always there), it's not clear if you're looking at this in isolation. Better to use .find(), even though that adds to the length.

Reply:
I think you worry too much here. Earlier I have this line:
`masterVassals[V2Tag1].insert(V2Tag2);`
It doesn't check if V2Tag1 is in map nor if there exists a set to store it, or if the value is actually stored. C++ handles all of this, which is the beauty of maps.

When iterating over a map with a const auto& I know the map is immutable, so no worries of changing anything. I'd like to reserve iterator-incrementing loops or .find() for instances where one has to alter the element within the map.

For this particular case, I think find() would be excessive and harder to read (at least to me).